src/Components/Toolbar/SaveCharacter/SaveCharacter.tsx

```diff
@@ -1,53 +1,50 @@
-import React from 'react';
-import { setDoc } from '@firebase/firestore';
-import { useContext, useState } from 'react';
+import React, { useContext, useState } from 'react';
+import { DocumentData, DocumentReference } from '@firebase/firestore';
 
 import Button from 'react-bootstrap/Button';
 
-import { UserContext } from '../../context';
-import { addNewCharacterForUser, getUsersSavedCharactersList } from '../Firebase/firebase.utils';
-import { NewCharacterStatsObj } from '../../../types';
+import { UserContext } from '../../../context';
+import {
+  addNewCharacterForUser,
+  createCharacterDocument,
+  getUsersSavedCharacterList
+} from '../../Firebase/firebase.utils';
+import { NewCharacterStatsObj } from '../../../../types';
 import { setId } from 'features/characterSlice';
-import { useAppDispatch, useAppSelector } from 'Components/CustomHooks/reduxHooks';
+import { useAppDispatch, useAppSelector } from 'features/reduxHooks';
 
 interface Props {
-  setShowSaveAlert: React.Dispatch<React.SetStateAction<boolean>>;
+  setShowAlert: React.Dispatch<React.SetStateAction<string | null>>;
 }
 
-const SaveCharacter = ({ setShowSaveAlert }: Props) => {
+const SaveCharacter = ({ setShowAlert }: Props) => {
   const dispatch = useAppDispatch();
   const userContext = useContext(UserContext);
   const user = userContext?.user;
   const userId = user ? user.uid : '';
   const characterName = useAppSelector((state) => state.character.name);
   const selectedAdvantages = useAppSelector((state) => state.character.advantages);
   const selectedDisadvantages = useAppSelector((state) => state.character.disadvantages);
-
   const [isSaving, setIsSaving] = useState(false);
 
   const checkSaveRequirements = async () => {
     const checkIfDuplicate = async () => {
       if (user) {
-        const characterList = await getUsersSavedCharactersList(userId);
+        const characterList = await getUsersSavedCharacterList(userId);
```

Comment:
Theres a lot of returns here returning nothing which just ends up being undefined, then also a lot of the functions are returning multiple types. you wanna try and return a single type in most cases and avoid returning undefined. in this case checkIfDuplicate should be just a Promise<bool>. you can also reduce the amount of promises being passes around by passing in values. then the duplicate checking is just a bool. and your awaiting can be done outside the function

eg
```ts
const characterList = await getUsersSavedCharacterList(userId);
isDuplicate(characterList) // only true of false non async
```

Reply:
Some good lessons here.
I have refactored with them in mind.

Let me know if you think I can improve it more.